beacon-chain/p2p/discovery.go

```diff
@@ -419,45 +439,143 @@ func parseGenericAddrs(addrs []string) (enodeString, multiAddrString []string) {
 }
 
 func convertToMultiAddr(nodes []*enode.Node) []ma.Multiaddr {
-	var multiAddrs []ma.Multiaddr
+	// Expect each node to have a TCP and a QUIC address.
+	multiAddrs := make([]ma.Multiaddr, 0, 2*len(nodes))
+
 	for _, node := range nodes {
-		// ignore nodes with no ip address stored
+		// Skip nodes with no ip address stored.
 		if node.IP() == nil {
 			continue
 		}
-		multiAddr, err := convertToSingleMultiAddr(node)
+
+		// Get up to two multiaddrs (TCP and QUIC) for each node.
+		nodeMultiAddrs, err := convertToMultiAddrs(node)
 		if err != nil {
-			log.WithError(err).Error("Could not convert to multiAddr")
+			log.WithError(err).Errorf("Could not convert to multiAddr node %s", node)
 			continue
 		}
-		multiAddrs = append(multiAddrs, multiAddr)
+
+		multiAddrs = append(multiAddrs, nodeMultiAddrs...)
 	}
+
 	return multiAddrs
 }
 
-func convertToAddrInfo(node *enode.Node) (*peer.AddrInfo, ma.Multiaddr, error) {
-	multiAddr, err := convertToSingleMultiAddr(node)
+func convertToAddrInfo(node *enode.Node) (*peer.AddrInfo, []ma.Multiaddr, error) {
+	multiAddrs, err := convertToMultiAddrs(node)
 	if err != nil {
 		return nil, nil, err
 	}
-	info, err := peer.AddrInfoFromP2pAddr(multiAddr)
+
+	if len(multiAddrs) == 0 {
+		return nil, nil, nil
+	}
+
+	infos, err := peer.AddrInfosFromP2pAddrs(multiAddrs...)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, errors.Wrapf(err, "could not convert to peer info: %v", multiAddrs)
+	}
+
+	if len(infos) > 1 {
```

Comment:
This seems wrong, what happens if we have 2 multiaddrs returned from `convertToMultiAddrs` . This error would always be triggered.

Reply:
`peer.AddrInfosFromP2pAddrs` groups input `maddrs` by node.

If `multiaddrs` contains 5 `multiaddr`s, but belonging to only 2 nodes, (for example 3 `multiaddr`s for one node and 2 `multiaddr`s for the other node), then `len(infos) == 2`  and:
- `len(infos[0].Addrs == 3`
- `len(infos[1].Addrs == 2`

Reply:
I also added a check in `func (s *Service) filterPeer(node *enode.Node) bool`.

In this function, even if the only way to have `peerData == nil` is to have at the same time `multiAddrs == nil`, it's better to check `peerData` nilness explicitely.

Reply:
Fixed in e9d80e2dfa2f629e8629ce0f44667f70e11b3648.